### PR TITLE
Add Copy Vanilla CSS

### LINKED ISSUE #4: Add a 'Copy vanilla CSS' option
As [stated here](https://dev.to/jcalixte/comment/16jb2), it would be cool to have a button which copies the vanilla CSS code for a gradient too.

### PR DESCRIPTION
**Pull request by Drarig29**
### Context
Closes #4 

### Changes proposed in this pull request
- _Add `package-lock.json` to `.gitignore` because this project uses Yarn_
- Add a button to copy the vanilla `linear-gradient()`
  - Add a `vanilla` prop to keep one `ClickCopy` component
  - Add a `value` prop which holds the value to be copied, which can either be Tailwind classes or a `linear-gradient()`

### Guidance to review

I have two concerns:
- Maybe change the buttons' labels to be more specific about what they are copying (e.g. "Copy Tailwind classes" and "Copy Vanilla gradient")
- Basically, **`getComputedStyle()` needs a reference to a DOM element**. Therefore I tried using `ref`, but it's populated after the first render, so the initial load of `ClickCopy` can't access the DOM element. My workaround was to use a prop that I mutate in `mounted()` and `updated()`, but it's bad practice.

I just discovered Vue.js and tried to do my best, but because of the latter point there are warnings in the console.

I'm waiting to see what you will suggest to me 😀

### CODE REVIEW

File: src/components/ClickCopy.vue [hunk 1]
```diff
@@ -25,6 +25,14 @@ export default {
...
     gradient: String,
     name: String,
+    vanilla: Boolean,
+    value: String,
+  },
+  mounted() {
+    this.value = this.vanilla ? getComputedStyle(this.$refs.gradientUnderline).getPropertyValue('background-image') : this.gradient
+  },
+  updated() {
+    this.value = this.vanilla ? getComputedStyle(this.$refs.gradientUnderline).getPropertyValue('background-image') : this.gradient
```
Comment (jcalixte) [inline src/components/ClickCopy.vue:35]: Great use of getComputedStyle!
First I think what you want is having `value` as a `data` instead of a prop (in VueJS, props are inputs, not internal state), but maybe the best option is to have a `vanillaStyle` [computed property](https://vuejs.org/v2/guide/computed.html).

```suggestion
  },
  computed: {
    vanillaStyle() {
       return getComputedStyle(this.$refs.gradientUnderline).getPropertyValue('background-image')
    }
```

Comment (Drarig29) [reply src/components/ClickCopy.vue:35]: Okay, but this doesn't solve the fact that `this.$refs.gradientUnderline` is undefined during the first render of the component. So `getComputedStyle()` breaks and the component isn't rendered.

The problem is that `v-clipboard:copy` evaluates the value during the first render, so I can't use the `ref` yet. I don't know if there is another way to do it.

Doing without the `ref` is hard (with `document` DOM methods) because it's hard to find the good query.

Maybe the only solution is to move the logic up in the parent component (i.e. `Gradient`)...

Comment (jcalixte) [reply src/components/ClickCopy.vue:35]: What do you think of:
```js
if (!this.refs.gradientUnderline) {
  return null
}
return ...
```

Comment (Drarig29) [reply src/components/ClickCopy.vue:35]: Then it won't copy anything and you'll need to select other directions to re-render the component so as to have something to copy... It's not a solution I think 😁

Comment (jcalixte) [reply src/components/ClickCopy.vue:35]: Damn it...

Comment (jcalixte) [reply src/components/ClickCopy.vue:35]: I found it, `refs` is not reactive (https://stackoverflow.com/a/43531779), but, `v-clipboard:copy` can have a computed value in it.

Here a code that works:

```js
  data() {
    return {
      mounted: false,
    };
  },
  props: {
    gradient: String,
    name: String,
    vanilla: Boolean,
  },
  mounted() {
    this.mounted = true;
  },
  computed: {
    value() {
      if (!this.mounted) {
        return null;
      }
      if (this.vanilla) {
        return getComputedStyle(this.$refs.gradientUnderline).getPropertyValue(
          "background-image"
        );
      }
      return this.gradient;
    },
  },
```

Comment (Drarig29) [reply src/components/ClickCopy.vue]: Okay, thanks. I just implemented it and it works well!

File: src/components/ClickCopy.vue [hunk 2]
```diff
@@ -25,6 +25,26 @@ export default {
...
+  computed: {
+    value() {
+      if (!this.mounted) return null
+      
+      if (this.vanilla) {
+        return getComputedStyle(this.$refs.gradientUnderline).getPropertyValue('background-image')
+      }
+
+      return this.gradient
+    }
```
Comment (markmead) [inline src/components/ClickCopy.vue:47]: All this logic in the components that's sole purpose is to copy a value to the clipboard feels overkill.

I also think if we were to add this feature, I would want it separate from the current button style and position, so a new component is an option.

File: src/components/ClickCopy.vue [hunk 3]
```diff
@@ -5,16 +5,16 @@
...
         v-clipboard:error="onError"
         :data-gradient="name"
       >
-        Copy CSS
+        Copy <span v-if="vanilla">Vanilla</span> CSS
       </button>
       <span
         class="inline-block w-6 h-1 mt-1 duration-100 ease-in-out rounded-full transition-size group-hover:w-full"
-        :class="gradient"
+        :class="gradient" ref="gradientUnderline"
```
Comment (markmead) [inline src/components/ClickCopy.vue:17]: I would use the main "gradient" part as the `ref`.

**Review by markmead — COMMENTED** [review]
The logic in the "ClickCopy" component can be replaced with a `mounted` call on the "Gradient" component. It's cleaner and keeps the logic in "ClickCopy" clean.

```vue
mounted() {
    this.$nextTick(() => (this.css = window.getComputedStyle(this.$refs.gradient).getPropertyValue('background-image')))
  }
```

Where we update a `css` data value that we can then pass to a "Click to Copy" component.

**Comment by markmead** [comment]
If we are to add this button we need to think about the wording and placement of the button. It's a Tailwind collection, so do we really need it? And if we do have it where should the button be placed?

Thoughts @jordihales

File: src/components/ClickCopy.vue (hunk 3 above)
Comment (Drarig29) [reply src/components/ClickCopy.vue:17]: What do you mean by "main gradient part"? I don't understand the suggestion.

Comment (markmead) [reply src/components/ClickCopy.vue:17]: Sorry, that was very vague 😅

[This line here](https://github.com/jordihales/hypercolor/blob/master/src/components/Gradient.vue#L3)

**Comment by Drarig29** [comment]
Those gradients are nice so I think it would be lame to force people using Tailwind CSS. Personally, if I needed only one gradient for a project, I would not install the library just for that...

For the matter of the button placement, maybe you could add something like this (something kind of hidden):

image

I could change the placement if you want. The PR is mostly to make a POC of the feature.

**Comment by markmead** [comment]
> Those gradients are nice so I think it would be lame to force people using Tailwind CSS. Personally, if I needed only one gradient for a project, I would not install the library just for that...

I understand your point, but this was directly built for Tailwind users.

> I could change the placement if you want. The PR is mostly to make a POC of the feature.

The placement, wording, and style of the button would need to be changed. I'm not convinced a dropdown is the way forward as it slows down the current process. As this was built with Tailwind users, every addition has to be complementary to a Tailwind users process on the site. 

As this is a POC I think we should park it for now, if more users start requesting this functionality we can come back to this PR.
